File: Problema5/challenge.java
/**
 * @file challenge.java
 * @author Alan Davide Bovo (alandavide.bovo.stud@ispascalcomandini.it)
 * @date 2024-01-06
 * Compile with: javac Problema5/challenge.java
 * Run with: java Problema5.challenge
 */
package Problema5;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Random;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public class challenge {
    private static byte[] calculateSHA256(String input) { // Funzione per calcolare l'SHA-256
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(input.getBytes());
            return hash;
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
            return null;
        }
    }
    public static void main(String[] args) {
        String flag = ""; // Stringa per contenere la flag
        try {
            flag = new String(Files.readAllBytes(Paths.get("flag.txt")));   // Leggo il flag dal file
        } catch (Exception e) {   // Errori vari
            System.out.println(e.getMessage()); // Stampo l'errore
            System.exit(1);                     // Esco con errore
        }

        Random rnd = new Random();                                              // Inizializzo il generatore di numeri casuali
        byte[] hash = calculateSHA256(String.format("%d", rnd.nextInt(256)));   // Come in C#
        System.out.print("The encrypted flag is: ");                            // Stampo il messaggio
        for(int i=0; i<hash.length; i++) {
            System.out.printf("%02x", (hash[i] ^ flag.codePointAt(i)) & 0xFF);  // Stampo il risultato in esadecimale
        }
    }
}

File: Problema5/challenge.kt
/**
 * @file challenge.kt
 * @author Alan Davide Bovo (alandavide.bovo.stud@ispascalcomandini.it)
 * @date 2024-01-06
 * Compile with: kotlinc challenge.kt -include-runtime -d challenge.jar
 * Run with: java -jar challenge.jar
 */
import java.io.File
import java.util.Random
import java.security.MessageDigest

fun calculateSHA256(input: String): ByteArray { // Funzione per calcolare l'hash SHA256 di una stringa
    val md = MessageDigest.getInstance("SHA-256")
    val hashBytes = md.digest(input.toByteArray())
    return hashBytes
}

fun main() {
    val flagFile = File("flag.txt") // Apro il file flag.txt
    val flag = flagFile.readText()  // Leggo il contenuto del file flag.txt
    
    val rnd = Random()              // Creo un oggetto Random
    val hash = calculateSHA256("${rnd.nextInt(256)}")   // Calcolo l'hash SHA256 di un numero casuale
    print("The encrypted flag is: ")                    // Stampo il messaggio
    for(i in 0..31) {
        print("%02x".format((hash[i].toInt() xor flag[i].code) and 0xFF)) // Cripto la flag con l'hash calcolato
    }
}